File: frontend/src/services/categoryService.ts
import api from "./api";

export interface Category {
  category_id: number;
  category_name: string;
  description?: string | null;
}

export interface CategoryCreate {
  category_name: string;
  description?: string;
}

export interface CategoryUpdate {
  category_name: string;
  description?: string;
}

/*
 * Get all categories
 */
export const getCategories = async (): Promise<Category[]> => {
  const response = await api.get<Category[]>("/categories/");
  return response.data;
};

/*
 * Get category by ID
 */
export const getCategoryById = async (
  categoryId: number
): Promise<Category> => {
  const response = await api.get<Category>(
    `/categories/${categoryId}`
  );

  return response.data;
};

/*
 * Create category
 */
export const createCategory = async (
  category: CategoryCreate
): Promise<Category> => {
  const response = await api.post<Category>(
    "/categories/",
    category
  );

  return response.data;
};

/*
 * Update category
 */
export const updateCategory = async (
  categoryId: number,
  category: CategoryUpdate
): Promise<Category> => {
  const response = await api.put<Category>(
    `/categories/${categoryId}`,
    category
  );

  return response.data;
};

/*
 * Delete category
 */
export const deleteCategory = async (
  categoryId: number
): Promise<void> => {
  await api.delete(`/categories/${categoryId}`);
};

File: backend/app/services/category_service.py
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.models.category import Category
from app.schemas.category_schema import (
    CategoryCreate,
    CategoryUpdate,
)


def create_category(db: Session, category: CategoryCreate):

    existing = (
        db.query(Category)
        .filter(Category.category_name == category.category_name)
        .first()
    )

    if existing:
        raise ValueError("Category already exists.")

    new_category = Category(
        category_name=category.category_name,
        description=category.description,
    )

    db.add(new_category)
    db.commit()
    db.refresh(new_category)

    return new_category


def get_all_categories(db: Session):
    return db.query(Category).all()


def get_category_by_id(db: Session, category_id: int):
    return (
        db.query(Category)
        .filter(Category.category_id == category_id)
        .first()
    )


def update_category(
    db: Session,
    category_id: int,
    category: CategoryUpdate,
):
    existing = get_category_by_id(db, category_id)

    if not existing:
        raise ValueError("Category not found.")

    duplicate = (
        db.query(Category)
        .filter(
            Category.category_name == category.category_name,
            Category.category_id != category_id,
        )
        .first()
    )

    if duplicate:
        raise ValueError("Category already exists.")

    existing.category_name = category.category_name
    existing.description = category.description

    db.commit()
    db.refresh(existing)

    return existing


def delete_category(db: Session, category_id: int):

    category = get_category_by_id(db, category_id)

    if not category:
        raise ValueError("Category not found.")

    try:
        db.delete(category)
        db.commit()

        return {
            "message": "Category deleted successfully."
        }

    except IntegrityError:
        db.rollback()

        raise ValueError(
            "Cannot delete category because it is being used by another record."
        )